File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_TakeOver.java
package com.broll.gainea.server.core.cards.impl.play;

import com.broll.gainea.server.core.cards.Card;
import com.broll.gainea.server.core.objects.Unit;
import com.broll.gainea.server.core.utils.PlayerUtils;
import com.broll.gainea.server.core.utils.SelectionUtils;
import com.broll.gainea.server.core.utils.UnitControl;
import com.google.common.collect.Lists;

import java.util.List;
import java.util.stream.Collectors;

public class C_TakeOver extends Card {
    public C_TakeOver() {
        super(17, "Treueschwur", "Übernehmt die Einheit eines anderen Spielers (Ausser Feldherr)" +
                " und teleportiert sie zu eurem Feldherr.");
        setDrawChance(0.5f);
    }

    @Override
    public boolean isPlayable() {
        return PlayerUtils.isCommanderAlive(owner) && !getTargets().isEmpty();
    }

    private List<Unit> getTargets() {
        return PlayerUtils.getOtherPlayers(game, owner).flatMap(it -> it.getUnits().stream()).filter(it -> !PlayerUtils.isCommander(it)).collect(Collectors.toList());
    }

    @Override
    protected void play() {
        Unit unit = SelectionUtils.selectUnit(game, owner, "Welche Einheit soll übernommen werden?", getTargets());
        UnitControl.recruit(game, owner, Lists.newArrayList(unit), PlayerUtils.getCommander(owner).get().getLocation());
    }
}

File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_TakeOver.kt
package com.broll.gainea.server.core.cards.impl.play

import com.broll.gainea.server.core.cards.Card
import com.broll.gainea.server.core.objects.Unit
import com.broll.gainea.server.core.utils.UnitControl.recruit
import com.broll.gainea.server.core.utils.getCommander
import com.broll.gainea.server.core.utils.getOtherPlayers
import com.broll.gainea.server.core.utils.isCommander
import com.broll.gainea.server.core.utils.isCommanderAlive
import com.broll.gainea.server.core.utils.selectUnit

class C_TakeOver : Card(17, "Treueschwur", "Übernehmt die Einheit eines anderen Spielers (Ausser Feldherr)" +
        " und teleportiert sie zu eurem Feldherr.") {
    init {
        drawChance = 0.5f
    }

    override val isPlayable: Boolean
        get() = owner.isCommanderAlive() && targets.isNotEmpty()
    private val targets: List<Unit>
        get() = game.getOtherPlayers(owner).flatMap { it.units }.filter { !it.isCommander() }

    override fun play() {
        val unit = game.selectUnit(owner, "Welche Einheit soll übernommen werden?", targets)!!
        game.recruit(owner, listOf(unit), owner.getCommander()?.location)
    }
}
